android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java

```diff
@@ -388,116 +427,292 @@ private void reLayout(View view) {
+            super(allocator,
+                    minBufferMs,
+                    maxBufferMs,
+                    bufferForPlaybackMs,
+                    bufferForPlaybackAfterRebufferMs,
+                    targetBufferBytes,
+                    prioritizeTimeOverSizeThresholds,
+                    backBufferDurationMs,
+                    retainBackBufferFromKeyframe);
+            runtime = Runtime.getRuntime();
+            ActivityManager activityManager = (ActivityManager) themedReactContext.getSystemService(themedReactContext.ACTIVITY_SERVICE);
+            availableHeapInBytes = (int) Math.floor(activityManager.getMemoryClass() * maxHeapAllocationPercent * 1024 * 1024);
+        }
+
+        @Override
+        public boolean shouldContinueLoading(long playbackPositionUs, long bufferedDurationUs, float playbackSpeed) {
+            if (ReactExoplayerView.this.disableBuffering) {
+                return false;
+            }
+            int loadedBytes = getAllocator().getTotalBytesAllocated();
+            boolean isHeapReached = availableHeapInBytes > 0 && loadedBytes >= availableHeapInBytes;
+            if (isHeapReached) {
+                return false;
+            }
+            long usedMemory = runtime.totalMemory() - runtime.freeMemory();
+            long freeMemory = runtime.maxMemory() - usedMemory;
+            long reserveMemory = (long)minBufferMemoryReservePercent * runtime.maxMemory();
+            long bufferedMs = bufferedDurationUs / (long)1000;
+            if (reserveMemory > freeMemory && bufferedMs > 2000) {
+                // We don't have enough memory in reserve so we stop buffering to allow other components to use it instead
+                return false;
+            }
+            if (runtime.freeMemory() == 0) {
+                Log.w("ExoPlayer Warning", "Free memory reached 0, forcing garbage collection");
+                runtime.gc();
+                return false;
+            }
+            return super.shouldContinueLoading(playbackPositionUs, bufferedDurationUs, playbackSpeed);
+        }
+    }
+
+    private void startBufferCheckTimer() {
+        SimpleExoPlayer player = this.player;
+        VideoEventEmitter eventEmitter = this.eventEmitter;
+        Handler mainHandler = this.mainHandler;
+
+        if (this.bufferCheckTimer != null) {
+            this.stopBufferCheckTimer();
+        }
+
+        this.bufferCheckTimer = new Timer();
+        TimerTask bufferCheckTimerTask = new TimerTask() {
+            @Override
+            public void run() {
+                if (mainHandler != null) {
+                    mainHandler.post(new Runnable() {
+                        public void run() {
+                            if (player != null) {
+                                double bufferedDuration = (double) (player.getBufferedPercentage() * player.getDuration() / 100);
+                                eventEmitter.bufferProgress(0d, bufferedDuration);
```

Comment:
I don't understand why you need this new event ?
The onProgress event should be enough.
It may just need documentation clarification

Reply:
This was meant to be expanded with the buffer start being reported exactly (right now it's always 0, but it should be read in to represent the actual start position of the buffer). This meant to be slightly different than `onProgress` which just shows the buffered duration, this currently doesn't track the back buffer length, but it needs to be expanded in the future to show the start and end of buffer so that you know exactly where is the buffered area when using back buffer, since the playable/seekable duration in onProgress event has the total time which doesn't really tell where the buffer starts or ends when using both forward and back buffer.

Reply:
I created and issue for this https://github.com/react-native-video/react-native-video/issues/2695

Reply:
OK, I understand you need additionnal informations on available buffer, but it should be added to onProgress event.
the new bufferProgress has 2 paramters: 0 Ok for futur use, so not needed now.
bufferedDuration which is exactly equals to bufferedDuration in onProgress.
The only difference is the type you use double instead of long.

Another difference is the frequency. If another issue for you is linked to the the missing report when playback is pause, I invite you to check following PR: https://github.com/react-native-video/react-native-video/pull/2664

For me this new event doesn't bring new infomation for application...

Reply:
Other points are critical for me, but I really think this event shall be removed as it is duplicated from onProgress
It brings complexity are will make hard to maintain...
Can you please remove it and use onProgress in your app ? 
If there is issue or missing information in onProgress, let's fix it !

Reply:
I see that we added this so that we get buffer progress events even when paused. On progress doesn't seem like the best place to put this interval based reporting. I'm not sure if we want to keep receiving progress events when buffering. I feel that having the buffer quickly accessible on progress if needed is a nice to have, but having an independent buffer reporting that isn't tied to video progress I think is best. Does it make sense why this was separated out?

Reply:
@armadilio3 I have the same issue that you. I need to know if video is real buffering. There are some cases that we loose connection and it stop onProgress and start buffering, but i dont know if player is real buffering, because only the onProgress event show me that.

Reply:
@armadilio3 This is the last point to handle I think.
I still don't understand why you need to duplicate this event as information is already available in onProgress event.
once again there is a bug in onProgress which is blocking even when player is paused, it has to be solved at least for live playback.
Duplicating event can overload the bridge, so having too much event is a bad idea.

The advatage of onProgress is that timeout can already be configured.
I also think having 1 event with multiple parameters is lighter than have 2 events in terms of CPU consumption